File: src/test/java/ee/carlrobert/codegpt/conversations/ConversationsStateTest.java
package ee.carlrobert.codegpt.conversations;

import static org.assertj.core.api.Assertions.assertThat;

import com.intellij.testFramework.fixtures.BasePlatformTestCase;
import ee.carlrobert.codegpt.conversations.message.Message;
import ee.carlrobert.codegpt.settings.service.ServiceType;
import ee.carlrobert.codegpt.settings.state.OpenAISettingsState;
import ee.carlrobert.codegpt.settings.state.SettingsState;
import ee.carlrobert.llm.client.openai.completion.chat.OpenAIChatCompletionModel;

public class ConversationsStateTest extends BasePlatformTestCase {

  public void testStartNewDefaultConversation() {
    var settings = SettingsState.getInstance();
    settings.setSelectedService(ServiceType.OPENAI);
    OpenAISettingsState.getInstance().setModel(OpenAIChatCompletionModel.GPT_3_5.getCode());

    var conversation = ConversationService.getInstance().startConversation();

    assertThat(conversation).isEqualTo(ConversationsState.getCurrentConversation());
    assertThat(conversation)
        .extracting("clientCode", "model")
        .containsExactly("chat.completion", "gpt-3.5-turbo");
  }

  public void testSaveConversation() {
    var service = ConversationService.getInstance();
    var conversation = service.createConversation("chat.completion");
    service.addConversation(conversation);
    var message = new Message("TEST_PROMPT");
    message.setResponse("TEST_RESPONSE");
    conversation.addMessage(message);

    service.saveConversation(conversation);

    var currentConversation = ConversationsState.getCurrentConversation();
    assertThat(currentConversation).isNotNull();
    assertThat(currentConversation.getMessages())
        .flatExtracting("prompt", "response")
        .containsExactly("TEST_PROMPT", "TEST_RESPONSE");
  }

  public void testGetPreviousConversation() {
    var service = ConversationService.getInstance();
    var firstConversation = service.startConversation();
    service.startConversation();

    var previousConversation = service.getPreviousConversation();

    assertThat(previousConversation.isPresent()).isTrue();
    assertThat(previousConversation.get()).isEqualTo(firstConversation);
  }

  public void testGetNextConversation() {
    var service = ConversationService.getInstance();
    var firstConversation = service.startConversation();
    var secondConversation = service.startConversation();
    ConversationsState.getInstance().setCurrentConversation(firstConversation);

    var nextConversation = service.getNextConversation();

    assertThat(nextConversation.isPresent()).isTrue();
    assertThat(nextConversation.get()).isEqualTo(secondConversation);
  }

  public void testDeleteSelectedConversation() {
    var service = ConversationService.getInstance();
    var firstConversation = service.startConversation();
    service.startConversation();

    service.deleteSelectedConversation();

    assertThat(ConversationsState.getCurrentConversation()).isEqualTo(firstConversation);
    assertThat(service.getSortedConversations().size()).isEqualTo(1);
    assertThat(service.getSortedConversations())
        .extracting("id")
        .containsExactly(firstConversation.getId());
  }

  public void testClearAllConversations() {
    var service = ConversationService.getInstance();
    service.startConversation();
    service.startConversation();

    service.clearAll();

    assertThat(ConversationsState.getCurrentConversation()).isNull();
    assertThat(service.getSortedConversations().size()).isEqualTo(0);
  }
}

File: src/test/kotlin/ee/carlrobert/codegpt/conversations/ConversationsStateTest.kt
package ee.carlrobert.codegpt.conversations

import com.intellij.testFramework.fixtures.BasePlatformTestCase
import ee.carlrobert.codegpt.conversations.message.Message
import ee.carlrobert.codegpt.settings.GeneralSettings
import ee.carlrobert.codegpt.settings.service.ServiceType
import ee.carlrobert.codegpt.settings.service.openai.OpenAISettings
import ee.carlrobert.llm.client.openai.completion.OpenAIChatCompletionModel
import org.assertj.core.api.Assertions.assertThat

class ConversationsStateTest : BasePlatformTestCase() {

  fun testStartNewDefaultConversation() {
    GeneralSettings.getCurrentState().selectedService = ServiceType.OPENAI
    OpenAISettings.getCurrentState().model = OpenAIChatCompletionModel.GPT_3_5.code

    val conversation = ConversationService.getInstance().startConversation()

    assertThat(conversation).isEqualTo(ConversationsState.getCurrentConversation())
    assertThat(conversation)
      .extracting("clientCode", "model")
      .containsExactly("chat.completion", "gpt-3.5-turbo")
  }

  fun testSaveConversation() {
    val service = ConversationService.getInstance()
    val conversation = service.createConversation("chat.completion")
    service.addConversation(conversation)
    val message = Message("TEST_PROMPT")
    message.response = "TEST_RESPONSE"
    conversation.addMessage(message)

    service.saveConversation(conversation)

    val currentConversation = ConversationsState.getCurrentConversation()
    assertThat(currentConversation).isNotNull()
    assertThat(currentConversation!!.messages)
      .flatExtracting("prompt", "response")
      .containsExactly("TEST_PROMPT", "TEST_RESPONSE")
  }

  fun testGetPreviousConversation() {
    val service = ConversationService.getInstance()
    val firstConversation = service.startConversation()
    service.startConversation()

    val previousConversation = service.previousConversation

    assertThat(previousConversation.isPresent).isTrue()
    assertThat(previousConversation.get()).isEqualTo(firstConversation)
  }

  fun testGetNextConversation() {
    val service = ConversationService.getInstance()
    val firstConversation = service.startConversation()
    val secondConversation = service.startConversation()
    ConversationsState.getInstance().setCurrentConversation(firstConversation)

    val nextConversation = service.nextConversation

    assertThat(nextConversation.isPresent).isTrue()
    assertThat(nextConversation.get()).isEqualTo(secondConversation)
  }

  fun testDeleteSelectedConversation() {
    val service = ConversationService.getInstance()
    val firstConversation = service.startConversation()
    service.startConversation()

    service.deleteSelectedConversation()

    assertThat(ConversationsState.getCurrentConversation()).isEqualTo(firstConversation)
    assertThat(service.sortedConversations.size).isEqualTo(1)
    assertThat(service.sortedConversations)
      .extracting("id")
      .containsExactly(firstConversation.id)
  }

  fun testClearAllConversations() {
    val service = ConversationService.getInstance()
    service.startConversation()
    service.startConversation()

    service.clearAll()

    assertThat(ConversationsState.getCurrentConversation()).isNull()
    assertThat(service.sortedConversations.size).isEqualTo(0)
  }
}
